packages/@glimmer/interfaces/lib/program.d.ts

```diff
@@ -79,25 +82,48 @@ export type ConstantPool = unknown[];
 export interface CompileTimeConstants {
   value(value: unknown): number;
   array(values: unknown[] | readonly unknown[]): number;
-  serializable(value: unknown): number;
   toPool(): ConstantPool;
 }
 
 /**
- * In JIT mode, the constant pool is allowed to store arbitrary values,
- * which don't need to be serialized or transmitted over the wire.
+ * Resolution happens when components are first loaded, either via the resolver
+ * or via looking them up in template scope.
  */
-export interface CompileTimeLazyConstants extends CompileTimeConstants {
-  other(value: unknown): number;
+export interface ResolutionTimeConstants {
+  // TODO: The default template is unique per-program. This should likely belong
+  // in StdLib, but it's not really possible to thread it through that way
+  // currently.
+  defaultTemplate: Template;
+
+  helper(
+    owner: Owner | undefined,
+    definitionState: HelperDefinitionState,
+    resolvedName: string | null
+  ): number;
+
+  modifier(
+    owner: Owner | undefined,
+    definitionState: ModifierDefinitionState,
+    resolvedName: string | null
+  ): number;
+
+  component(
+    owner: Owner | undefined,
+    definitionState: ComponentDefinitionState
+  ): ComponentDefinition;
+
+  resolvedComponent(
```

Comment:
Essentially because in general, we want the resolved values to match the same values that would have been used in strict mode. For helpers/modifiers, we can do this 1-1, but for components when they are resolved we cannot, because we exposed some of the internal details, such as the fact that you can resolve the same component definition with different templates. So, we need to have a different type, and different logic for adding it to the constants pool.